control-plane/pkg/util/lock_map.go

```diff
@@ -23,23 +23,24 @@ import (
 )
 
 // LockMap provides a mechanism to get a lock for a given key.
-type LockMap interface {
-	GetLock(interface{}) *sync.Mutex
+type LockMap[K comparable] interface {
+	GetLock(K) *sync.Mutex
 }
 
-type expiringLockMap struct {
+type expiringLockMap[K comparable] struct {
 	mu     sync.Mutex
-	locks  map[interface{}]*sync.Mutex
-	access map[interface{}]time.Time
+	locks  map[K]*sync.Mutex
+	access map[K]time.Time
 
 	ttl time.Duration
 }
 
-// NewExpiringLockMap returns a new LockMap that removes entries after the given TTL.
-func NewExpiringLockMap(ctx context.Context, ttl time.Duration) LockMap {
-	lm := &expiringLockMap{
-		locks:  make(map[interface{}]*sync.Mutex),
-		access: make(map[interface{}]time.Time),
+// NewExpiringLockMap returns a new LockMap that removes entries after the given TTL. Timing of the removal is done
+// in a best-effort fashion, and is not guaranteed to be exact.
+func NewExpiringLockMap[K comparable](ctx context.Context, ttl time.Duration) LockMap[K] {
```

Comment:
Wouldn't be useful to abstract on the "value" as well.

I have already a use of this generic map in the prober cache

Reply:
The map in prober cache was specific to Status AFAI see, so I didn't use that.

When I started this `LockMap`, I was thinking about providing `Lock(key string)` and `Unlock(key string)` methods that were supposed to lock/unlock the mutexes. That didn't work...

Actually, currently this is a very generic map that can store anything.. I can work on this in a separate PR.